### PR TITLE
Add DNS post-push test

### PR DESCRIPTION
**Pull request by hh**
Allows for querying of dns servers / verification that zone configuration is published.

xref #151

### CODE REVIEW

File: dns/octodns-docker/check-zone.py [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+        source.populate(zone)
+
+    print('name,type,ttl,{},consistent'.format(','.join(args.server)))
+    resolvers = []
+    ip_addr_re = re.compile(r'^[\d\.]+$')
+    for server in args.server:
+        resolver = AsyncResolver(configure=False,
+                                 num_workers=int(args.num_workers))
+        if not ip_addr_re.match(server):
+            server = unicode(query(server, 'A')[0])
```
Comment (thockin) [inline dns/octodns-docker/check-zone.py]: is it worthwhile to re-check the IP regex?

File: dns/octodns-docker/check-zone.py [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+    except KeyError as e:
+        raise Exception('Unknown source: {}'.format(e.args[0]))
+
+    zone = Zone(args.zone, manager.configured_sub_zones(args.zone))
+    for source in sources:
+        source.populate(zone)
+
+    print('name,type,ttl,{},consistent'.format(','.join(args.server)))
+    resolvers = []
+    ip_addr_re = re.compile(r'^[\d\.]+$')
```
Comment (thockin) [inline dns/octodns-docker/check-zone.py]: Not exactly right, I guess you're just trying to tell names from IPs?  Maybe a comment why this is sloppy?

File: dns/octodns-docker/check-zone.py [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+    for server in args.server:
+        resolver = AsyncResolver(configure=False,
+                                 num_workers=int(args.num_workers))
+        if not ip_addr_re.match(server):
```
Comment (thockin) [inline dns/octodns-docker/check-zone.py]: minor, but I would do this first in the block

File: dns/octodns-docker/check-zone.py [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+            except NXDOMAIN:
+                answers = ['*does not exist*']
+                dns_error = True
+            except Timeout:
+                answers = ['*timeout*']
+                dns_error = True
+            except (NoAnswer, NoNameservers):
+                # unsure why NS records come back: *** NoAnswer:
+                # The DNS response does not contain an answer to the question: X. IN NS
+                # However all other records wo can compare
```
Comment (thockin) [inline dns/octodns-docker/check-zone.py]: s/wo/we

File: dns/octodns-docker/check-zone.py [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+                if record._type == 'NS':
+                    continue
+                answers = ['*no answer*']
+                dns_error = True
+            stdout.write(' '.join(answers))
+            #depending on the type, let's get a list of all values
+            if hasattr(record,'values'):
+                values = record.values
```
Comment (thockin) [inline dns/octodns-docker/check-zone.py]: call this "expected" for clarity ?

File: dns/octodns-docker/check-zone.py [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+            else:
+                values = [record.value]
+            for value in values:
+                value = unicode(value).replace("'","")
+                if value in answers:
```
Comment (thockin) [inline dns/octodns-docker/check-zone.py]: This verifies that we have all the expected values but not that there are no unexpected values.  Compare length or do the inverse iteration?

File: dns/push.sh [hunk 7]
```diff
@@ -17,6 +17,32 @@ function push() {
...
+    gcloud dns managed-zones list --format=flattened --format='value(name)' --filter="dnsName=$1"
+}
+
+function get_zone_dns() {
+    gcloud dns managed-zones describe $1 --format='value(nameServers[0])'
+}
+
+function check_zone() {
+    DOMAIN=$1
+    GCLOUD_ZONE=$(domain_to_zone $DOMAIN)
```
[the remark this answers is not in the document]
Comment (hh) [reply dns/push.sh]: I refactored a bit to use the DNS servers configured in GCP.

File: dns/push.sh [hunk 8]
```diff
@@ -44,8 +70,12 @@ if [ $? != 0 ]; then
 fi
 echo "Canary push SUCCEEDED"
 
-# TODO: run test against canary
+check_zone "canary.k8s.io." || exit $?
+check_zone "canary.kubernetes.io." || exit $?
+echo "Canary test SUCCEEDED"
 
+# Exit for now until we are ready to push to prod
+exit 0
```
[the remark this answers is not in the document]
Comment (hh) [reply dns/push.sh]: Yep that should work!

File: dns/test.sh [hunk 9]
```diff
@@ -1,248 +1,50 @@
...
-  - ns-687.awsdns-21.net.
-# @michelle192837
-testgrid:
-  type: CNAME
-  value: redirect.k8s.io.
-# Project metrics.  Running in GKE (@spiffxp, @cjwagner)
-velodrome:
-  type: A
-  value: 104.197.200.129
+function domain_to_zone() {
```
[the remark this answers is not in the document]
Comment (hh) [reply dns/test.sh]: I moved the logic into check-zone.sh

File: dns/octodns-docker/check-zone.py [hunk 10]
```diff
@@ -0,0 +1,133 @@
...
+
+def main():
+    parser = ArgumentParser(description=__doc__.split('\n')[1])
+
+    parser.add_argument('--config-file', required=True,
+                        help='The Manager configuration file to use')
+    parser.add_argument('--zone', required=True, help='Zone to dump')
+    parser.add_argument('--source', required=True, default=[], action='append',
+                        help='Source(s) to pull data from')
+    parser.add_argument('--num-workers', default=4,
```
[the remark this answers is not in the document]
Comment (hh) [reply dns/octodns-docker/check-zone.py]: We started from https://github.com/github/octodns/blob/master/octodns/cmds/report.py as a boiler plate. We refactored it this morning.

File: dns/push.sh [hunk 11]
```diff
@@ -44,8 +44,25 @@ if [ $? != 0 ]; then
 fi
 echo "Canary push SUCCEEDED"
 
-# TODO: run test against canary
+echo "Testing canary zones"
+./check-zone.sh "canary.k8s.io." > log.canary 2>&1
```
Comment (thockin) [inline dns/push.sh]: should `>>` to the log

File: dns/push.sh [hunk 12]
```diff
@@ -44,8 +44,25 @@ if [ $? != 0 ]; then
...
+if [ $? != 0 ]; then
+    echo "Canary test FAILED, halting; log follows:"
+    echo "========================================="
+    cat log.canary
+    exit 2
+fi
+./check-zone.sh "canary.kubernetes.io." > log.canary 2>&1
```
Comment (thockin) [inline dns/push.sh]: should `>>` to the log

File: dns/push.sh [hunk 13]
```diff
@@ -65,4 +83,21 @@ if [ $? != 0 ]; then
 fi
 echo "Prod push SUCCEEDED"
 
-# TODO: run test against prod
+echo "Testing prod zones"
+./check-zone.sh "k8s.io." > log.prod 2>&1
```
Comment (thockin) [inline dns/push.sh]: should `>>` to the log

File: dns/push.sh [hunk 14]
```diff
@@ -65,4 +83,21 @@ if [ $? != 0 ]; then
...
+if [ $? != 0 ]; then
+    echo "Canary test FAILED, halting; log follows:"
+    echo "========================================="
+    cat log.canary
+    exit 2
+fi
+./check-zone.sh "kubernetes.io." > log.prod 2>&1
```
Comment (thockin) [inline dns/push.sh]: should `>>` to the log

File: dns/push.sh (hunk 8 above)
[the remark this answers is not in the document]
Comment (thockin) [reply dns/push.sh]: nix this exit, then?

File: dns/push.sh [hunk 15]
```diff
@@ -65,4 +83,21 @@ if [ $? != 0 ]; then
...
+if [ $? != 0 ]; then
+    echo "Canary test FAILED, halting; log follows:"
+    echo "========================================="
+    cat log.canary
+    exit 2
+fi
+echo "Canary test SUCCEEDED"
```
Comment (thockin) [inline dns/push.sh]: s/Canary/Prod

File: dns/push.sh [hunk 16]
```diff
@@ -65,4 +83,21 @@ if [ $? != 0 ]; then
...
+./check-zone.sh "k8s.io." > log.prod 2>&1
+if [ $? != 0 ]; then
+    echo "Canary test FAILED, halting; log follows:"
+    echo "========================================="
+    cat log.canary
+    exit 2
+fi
+./check-zone.sh "kubernetes.io." > log.prod 2>&1
+if [ $? != 0 ]; then
+    echo "Canary test FAILED, halting; log follows:"
```
Comment (thockin) [inline dns/push.sh]: s/Canary/Prod

File: dns/push.sh [hunk 17]
```diff
@@ -65,4 +83,21 @@ if [ $? != 0 ]; then
 fi
 echo "Prod push SUCCEEDED"
 
-# TODO: run test against prod
+echo "Testing prod zones"
+./check-zone.sh "k8s.io." > log.prod 2>&1
+if [ $? != 0 ]; then
+    echo "Canary test FAILED, halting; log follows:"
```
Comment (thockin) [inline dns/push.sh]: s/Canary/Prod

File: dns/push.sh [hunk 18]
```diff
@@ -65,4 +83,21 @@ if [ $? != 0 ]; then
...
+    echo "========================================="
+    cat log.canary
```
Comment (thockin) [inline dns/push.sh]: s/canary/prod

File: dns/push.sh [hunk 19]
```diff
@@ -65,4 +83,21 @@ if [ $? != 0 ]; then
...
+    exit 2
+fi
+./check-zone.sh "kubernetes.io." > log.prod 2>&1
+if [ $? != 0 ]; then
+    echo "Canary test FAILED, halting; log follows:"
+    echo "========================================="
+    cat log.canary
```
Comment (thockin) [inline dns/push.sh]: s/canary/prod

File: dns/push.sh [hunk 20]
```diff
@@ -1,4 +1,4 @@
-#!/bin/sh
+#!/bin/bash
```
Comment (thockin) [inline dns/push.sh]: are we actually using bash isms?

File: dns/check-zone.sh [hunk 21]
```diff
@@ -0,0 +1,19 @@
+#!/bin/bash
+#
+# check-zone.sh
+#
+# $1: "canary" or "prod"
```
Comment (thockin) [inline dns/check-zone.sh]: This doesn't match with usage in the caller

File: dns/octodns-docker/Dockerfile [hunk 22]
```diff
@@ -10,5 +10,7 @@ RUN pip install -r /requirements.txt && \
     echo "freezing deps, please make sure requirements.txt matches" && \
     pip freeze | tee /newrequirements.txt && \
     diff /requirements.txt /newrequirements.txt
-
+# RUN pip install ipython ipdb
```
Comment (thockin) [inline dns/octodns-docker/Dockerfile]: Why are these here but commented?  Do we need them or not?

File: dns/octodns-docker/check-zone.py [hunk 23]
```diff
@@ -0,0 +1,142 @@
...
+
+    args = parser.parse_args()
+    log = getLogger('report')
+
+    manager = Manager(args.config_file)
+    zone = Zone(args.zone, manager.configured_sub_zones(args.zone))
+    gcp = manager.providers['gcp']
+    project = gcp.gcloud_client.project
+    gcp.populate(zone)
+    #import ipdb ; ipdb.set_trace()
```
Comment (thockin) [inline dns/octodns-docker/check-zone.py]: Should we keep this or remove?

File: dns/octodns-docker/check-zone.py [hunk 24]
```diff
@@ -0,0 +1,142 @@
...
+
+    # Retrieve the DNS Servers directly from the GCP configuration
+    dns_servers = gcp.gcloud_zones[args.zone].name_servers
```
Comment (thockin) [inline dns/octodns-docker/check-zone.py]: At what step does this get populated?  I assume it has to reach out to GCP to find this...

File: dns/octodns-docker/check-zone.py (hunk 6 above)
Comment (hh) [reply dns/octodns-docker/check-zone.py]: The latest versions checks ensures no unexpected values, I went with inverse iteration.

File: dns/push.sh (hunk 8 above)
[the remark this answers is not in the document]
Comment (hh) [reply dns/push.sh]: Nixing.

File: dns/octodns-docker/check-zone.py (hunk 3 above)
Comment (hh) [reply dns/octodns-docker/check-zone.py]: done

File: dns/octodns-docker/check-zone.py (hunk 5 above)
Comment (hh) [reply dns/octodns-docker/check-zone.py]: We now setup _configured_values_ and _response_values_ for clarity.

File: dns/octodns-docker/check-zone.py (hunk 2 above)
Comment (hh) [reply dns/octodns-docker/check-zone.py]: I pulled this out into a function and made a cleaner regexp.
This was the approach used upstream. :)

File: dns/push.sh (hunk 15 above)
Comment (hh) [reply dns/push.sh]:
```
Dry-run to canary zones
Pushing to canary zones
Canary push SUCCEEDED
Testing canary zones
Canary test SUCCEEDED
Dry-run to prod zones
Pushing to prod zones
Prod push SUCCEEDED
This will likely fail the first time through... takes a bit to propogate
Testing prod zones
Canary test SUCCEEDED
Prod test SUCCEEDED
```

File: dns/push.sh (hunk 20 above)
Comment (hh) [reply dns/push.sh]: Removed 'function push ()' at https://github.com/kubernetes/k8s.io/commit/0d0051b5f809d97d45a19aaa33d95010df94aed5#diff-0a5093ca99f6474ed25e192ae7548a7dR7 replacing with 'push ()' at https://github.com/kubernetes/k8s.io/commit/5adb9a657e32e4ae1ed67ad88bc1c6faf128e265

Didn't look fully, but my dash shell didn't like it.

File: dns/octodns-docker/Dockerfile (hunk 22 above)
Comment (hh) [reply dns/octodns-docker/Dockerfile]: We don't need them, but they help with debugging. I turned them off/on several times.
I'll go ahead and remove them.

File: dns/octodns-docker/check-zone.py (hunk 23 above)
Comment (hh) [reply dns/octodns-docker/check-zone.py]: Removed. :+1:  :wastebasket:

File: dns/octodns-docker/check-zone.py (hunk 24 above)
Comment (hh) [reply dns/octodns-docker/check-zone.py]: I'm pretty sure it's at [gcp.populate](https://github.com/github/octodns/blob/master/octodns/provider/googlecloud.py#L193):

```python
    manager = Manager(args.config_file)                                                                                                                                                                            
    zone = Zone(args.zone, manager.configured_sub_zones(args.zone))                                                                                                                                                
    gcp = manager.providers['gcp']                                                                                                                                                                                 
    project = gcp.gcloud_client.project                                                                                                                                                                            
    gcp.populate(zone)    
```

**Comment by thockin** [comment]
It's fine to leave them commented out with a comment to enable them for
better debugging.

File: dns/octodns-docker/check-zone.py [hunk 25]
```diff
@@ -0,0 +1,198 @@
...
+        resolvers.append(resolver)
+    return resolvers
+
+def quote_cleanup(values):
+    """
+    Some cleanup for unicode u"'dns.domain.'" / u'"dns.domain."'
+    """
+    return [unicode(r).replace("'", "").replace('"', '') for r in values]
+
+def is_ip(dns_server):
```
Comment (thockin) [inline dns/octodns-docker/check-zone.py]: This is still not exactly right and more complex than needed.

What about:

```
import socket
try:
        socket.inet_pton(socket.AF_INET, address)
except socket.error:  # not a valid address
        return False
return True
```

File: dns/push.sh [hunk 26]
```diff
@@ -65,4 +81,19 @@ if [ $? != 0 ]; then
...
+./check-zone.sh "k8s.io." >> log.prod 2>&1
+if [ $? != 0 ]; then
+    echo "Prod test FAILED, halting; log follows:"
+    echo "========================================="
+    cat log.prod
+    exit 2
+fi
+./check-zone.sh "kubernetes.io." >> log.prod 2>&1
+if [ $? != 0 ]; then
+    echo "Production test FAILED, halting; log follows:"
```
Comment (thockin) [inline dns/push.sh]: s/Production/Prod (consistent)

**Comment by thockin** [comment]
/lgtm
/approve

Thanks @hh !!
